server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java

```diff
@@ -100,7 +106,17 @@ protected SearchPhase getNextPhase(SearchPhaseResults<CanMatchResponse> results,
         if (cardinality == 0) {
             // this is a special case where we have no hit but we need to get at least one search response in order
             // to produce a valid search result with all the aggs etc.
-            possibleMatches.set(0);
+            // Since it's possible that some of the shards that we're skipping are
+            // unavailable, we would try to query the node that at least has some
+            // shards available in order to produce a valid search result.
+            int shardIndexToQuery = 0;
+            for (int i = 0; i < shardsIts.size(); i++) {
+                if (shardsIts.get(i).size() > 0) {
+                    shardIndexToQuery = i;
+                    break;
+                }
+            }
+            possibleMatches.set(shardIndexToQuery);
```

Comment:
I understand that this is easy and fine for the existing can-match phase handling, but it could be bad for the coordinator can-match phase handling, since this could trigger an error. I wonder if we can improve this. Definitely OK to do in a follow-up, no need to block merging this PR on that.